src/NITTA/Model/ProcessorUnits/Comparator.hs

```diff
@@ -0,0 +1,306 @@
+
+    process = process_
+
+instance Connected (Comparator v x t) where
+    data Ports (Comparator v x t) = ComparePorts
+        { oePort :: SignalTag
+        , wrPort :: SignalTag
+        , opSelPort :: [SignalTag]
+        }
+        deriving (Show)
+
+supportedOpsNum :: Integer
+supportedOpsNum = fromIntegral $ length (dataTypeConstrs $ dataTypeOf F.CmpEq)
+selWidth = ceiling (logBase 2 (fromIntegral supportedOpsNum) :: Double) :: Int
+
+instance Controllable (Comparator v x t) where
+    data Instruction (Comparator v x t)
+        = Load F.CmpOp
+        | Out
+        deriving (Show)
+
+    data Microcode (Comparator v x t) = Microcode
+        { oe :: Bool
+        , wr :: Bool
+        , opSel :: Int
+        }
+        deriving (Show, Eq)
+
+    zipSignalTagsAndValues ComparePorts{..} Microcode{..} =
+        [ (oePort, Bool oe)
+        , (wrPort, Bool wr)
+        ]
+            ++ zipWith (\tag bit -> (tag, Bool bit)) opSelPort (bits opSel selWidth)
+        where
+            bits val localWidth = [testBit val (localWidth - idx - 1) | idx <- [0 .. localWidth - 1]]
+    usedPortTags ComparePorts{oePort, wrPort, opSelPort} = oePort : wrPort : opSelPort
+
+    takePortTags (oe : wr : xs) _ = ComparePorts oe wr sel
+        where
+            sel = take selWidth xs
+    takePortTags _ _ = error "can not take port tags, tags are over"
+
+instance Var v => Locks (Comparator v x t) v where
+    locks Comparator{remain, sources, targets} =
+        [ Lock{lockBy, locked}
+        | locked <- sources
+        , lockBy <- targets
+        ]
+            ++ [ Lock{lockBy, locked}
+               | locked <- concatMap (S.elems . variables) remain
+               , lockBy <- sources ++ targets
+               ]
+            ++ concatMap locks remain
+instance Default (Microcode (Comparator v x t)) where
+    def =
+        Microcode
+            { wr = False
+            , oe = False
+            , opSel = 0
+            }
+
+instance UnambiguouslyDecode (Comparator v x t) where
+    decodeInstruction Out = def{oe = True}
+    decodeInstruction (Load op) = case op of
+        F.CmpEq -> def{opSel = 0, wr = True}
+        F.CmpLt -> def{opSel = 1, wr = True}
+        F.CmpLte -> def{opSel = 2, wr = True}
+        F.CmpGt -> def{opSel = 3, wr = True}
+        F.CmpGte -> def{opSel = 4, wr = True}
+
+instance Default x => DefaultX (Comparator v x t) x
+
+instance Time t => Default (Comparator v x t) where
+    def = compare
+
+instance VarValTime v x t => EndpointProblem (Comparator v x t) v t where
+    endpointOptions pu@Comparator{targets = target : _} =
+        [EndpointSt (Target target) $ TimeConstraint at duration]
+        where
+            at = nextTick pu ... maxBound
+            duration = 1 ... maxBound
+    endpointOptions
+        pu@Comparator
+            { sources = _ : _
+            , currentWork = Just f
+            , process_
+            } = [EndpointSt (Source $ S.fromList (sources pu)) $ TimeConstraint at duration]
```

Comment:
Leave here TODO for the possibility of both input sequences: a, b and b, a. Or do it, it shouldn't be hard:

1. Write flipCompare Less -> More, Eq -> Eq...
2. Apply it, if the user starts from the second parameter.